File: Recrussion/csharp/Factorial.cs
using System;

namespace RecursionSolutions;

public static class Factorial
{
    public static long Solve(int n)
    {
        if (n < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(n), "n must be non-negative");
        }

        if (n <= 1)
        {
            return 1;
        }

        return n * Solve(n - 1);
    }
}

File: Recrussion/php/factorial.php

<?php

declare(strict_types=1);

function factorial(int $n): int
{
    if ($n < 0) {
        throw new InvalidArgumentException('n must be non-negative');
    }
    if ($n <= 1) {
        return 1;
    }

    return $n * factorial($n - 1);
}
